In [1]:
# Importing Libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
path=r'C:\Users\Usha Bharati\Downloads\02-2024 Instacart Basket Analysis'

In [3]:
# Importing Dataset
df_all_merged=pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'df_all_merged.pkl'))

## Q2 Addressing any PII data 

### The dataset contains Personal Identifiable Information(PII) namely first name, surname, age, family status, income. This information does not pertain to any person in real life. The data is completely fabricated for the purpose of learning in this course. 

## Q3 (a) Creating a “Region” column based on the “State” column from customers data set. 

## Creating 'Region' column using loc function

In [4]:
df_all_merged.loc[df_all_merged['state'].isin(['Maine','New Hampshire','Vermont','Massachusetts','Rhode Island','Connecticut','New York','Pennsylvania','New Jersey']),'region']='Northeast'

In [5]:
df_all_merged.loc[df_all_merged['state'].isin(['Wisconsin', 'Michigan', 'Illinois', 'Indiana', 'Ohio', 'North Dakota', 'South Dakota', 'Nebraska', 'Kansas', 'Minnesota', 'Iowa', 'Missouri']), 'region']='Midwest'

In [6]:
df_all_merged.loc[df_all_merged['state'].isin(['Delaware', 'Maryland', 'District of Columbia', 'Virginia', 'West Virginia', 'North Carolina', 'South Carolina', 'Georgia', 'Florida', 'Kentucky', 'Tennessee', 'Mississippi', 'Alabama', 'Oklahoma', 'Texas', 'Arkansas', 'Louisiana']), 'region'] = 'South'

In [7]:
df_all_merged.loc[df_all_merged['state'].isin(['Idaho','Montana','Wyoming','Nevada','Utah','Colorado','Arizona','New Mexico','Alaska','Washington','Oregon','California','Hawaii']),'region']='West'

In [8]:
df_all_merged['region'].value_counts(dropna=False)

region
South        10791885
West          8292913
Midwest       7597325
Northeast     5722736
Name: count, dtype: int64

## Q3 (b) Spending habits between the different U.S. regions

## Crosstab 'region' column with 'spending_flag'

In [9]:
spending_by_region = pd.crosstab(df_all_merged['region'], df_all_merged['spending_flag'], dropna = False).sort_values(by = 'High spender', ascending = False)

In [10]:
spending_by_region

spending_flag  High spender  Low spender
region                                  
South                209691     10582194
West                 160354      8132559
Midwest              155975      7441350
Northeast            108225      5614511

### 1.In the Southern U.S. region, there is a significant presence of High spenders, with 209,691 customers falling into this category. However, the majority of customers (10,582,194) are categorized as Low spenders.

### 2.Similarly, the Western and Midwestern regions exhibit substantial numbers of both High and Low spenders.

### 3.The Northeastern U.S. region has a comparatively lower count of High spenders (108,225) in comparison to the number of Low spenders (5,614,511).

## Q4 Creating an exclusion flag for low-activity customers and excluding them from the data

In [11]:
# Creating Exclusion Flag
df_all_merged['exclusion_flag'] = df_all_merged['max_order'] < 5

In [12]:
df_all_merged['exclusion_flag'].value_counts()

exclusion_flag
False    30964564
True      1440295
Name: count, dtype: int64

In [13]:
# Creating a new dataset which does not include data with max_order<5
df_filtered_data= df_all_merged[~df_all_merged['exclusion_flag']]

In [14]:
df_filtered_data.shape

(30964564, 34)

In [15]:
# Dataset with max_orders>5
sample_exclusion = df_all_merged[df_all_merged['exclusion_flag']]

In [16]:
# Exporting dataset as pickle format
df_filtered_data.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'df_filtered_data.pkl'))

In [17]:
# Exporting sample dataset in pickle format.
sample_exclusion.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'sample_exclusion.pkl'))